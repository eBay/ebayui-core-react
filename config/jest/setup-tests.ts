import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
