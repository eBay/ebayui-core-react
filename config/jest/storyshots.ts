import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
