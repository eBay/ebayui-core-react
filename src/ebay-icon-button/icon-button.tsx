import React, { ComponentProps, FC, RefObject } from 'react'
import classNames from 'classnames'
import { EbayIcon, Icon } from '../ebay-icon'
import { EbayBadge } from '../ebay-badge'
import { withForwardRef } from '../common/component-utils'

export type EbayIconButtonProps = {
    href?: string;
    icon: Icon;
    badgeNumber?: number;
    badgeAriaLabel?: string;
    transparent?: boolean;
    forwardedRef?: RefObject<HTMLAnchorElement & HTMLButtonElement>;
}

type HTMLButtonProps = ComponentProps<'button'>;
type HTMLAnchorProps = ComponentProps<'a'>;
type Props = EbayIconButtonProps & HTMLButtonProps & HTMLAnchorProps;

const EbayIconButton:FC<Props> = ({
    href,
    icon,
    badgeNumber,
    badgeAriaLabel,
    transparent,
    className: extraClasses,
    forwardedRef,
    ...rest
}) => {
    const classPrefix = href ? 'icon-link' : 'icon-btn'
    const className = classNames(
        extraClasses,
        classPrefix,
        {
            [`${classPrefix}--badged`]: badgeNumber,
            [`${classPrefix}--transparent`]: transparent
        }
    )
    const children = (
        <>
            <EbayIcon name={icon} />
            {badgeNumber && <EbayBadge type="icon" number={badgeNumber} aria-label={badgeAriaLabel} />}
        </>
    )

    return href ? (
        <a
            ref={forwardedRef}
            className={className}
            href={href}
            {...rest}
        >
            {children}
        </a>
    ) : (
        <button
            ref={forwardedRef}
            type="button"
            className={className}
            {...rest}
        >
            {children}
        </button>
    )
}

export default withForwardRef(EbayIconButton)
