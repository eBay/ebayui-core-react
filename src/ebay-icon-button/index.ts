export { default as EbayIconButton, type EbayIconButtonProps } from './icon-button'
