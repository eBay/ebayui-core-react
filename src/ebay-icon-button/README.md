# EbayIconButton

## Demo
[Storybook](https://opensource.ebay.com/ebayui-core-react/main/?path=/story/ebay-icon-button--default)

## Usage

### Import JS
```jsx harmony
import { EbayIconButton } from '@ebay/ui-core-react/ebay-icon-button'
```

### Import following styles from SKIN
```jsx harmony
import "@ebay/skin/icon-button"
```

### Or import styles using SCSS/CSS
```jsx harmony
import '@ebay/skin/icon-button.css'
```

### Icon button
```jsx harmony
<EbayIconButton icon="menu" badgeNumber={1} badgeAriaLabel="new feature available" />
```

## Attributes

Name | Type | Stateful | Required | Description | Data
--- | --- | --- | --- | --- | ---
`icon` | String | No | Yes | icon name
`href` | String | No | No | for link that looks like a button
`bageNumber` | Number | No | No | number on the badge
`bageAriaLabel` | String | No | No | aria label of the badge
`transparent` | Boolean | No | No | for transparent background
