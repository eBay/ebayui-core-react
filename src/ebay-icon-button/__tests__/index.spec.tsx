import React from 'react'
import { fireEvent, render } from '@testing-library/react';
import { initStoryshots } from '../../../config/jest/storyshots'
import { EbayIconButton } from '../../ebay-icon-button';

initStoryshots({
    config: ({ configure }) =>
        configure(() => {
            require('./index.stories')
        }, module)
})

describe('<EbayIconButton>', () => {
    describe('on passing a ref', () => {
        it('should set the ref to button element', () => {
            const ref = React.createRef() as any
            render(<EbayIconButton ref={ref} icon="menu" />)

            expect(ref.current.tagName).toBe('BUTTON')
        })
    })
    describe('on button click', () => {
        it('should fire onClick event', () => {
            const spy = jest.fn()
            const wrapper = render(
                <EbayIconButton onClick={spy} icon="add" />
            )
            fireEvent.click(wrapper.getByRole('button'))

            expect(spy).toBeCalled()
        })
    })
})
