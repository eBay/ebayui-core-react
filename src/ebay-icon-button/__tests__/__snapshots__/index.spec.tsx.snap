// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-icon-button Default 1`] = `
<DocumentFragment>
  <p>
    <button
      aria-label="Menu"
      class="icon-btn"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="icon icon--menu"
        focusable="false"
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-menu"
        />
      </svg>
    </button>
  </p>
  <p>
    <a
      aria-label="Settings"
      class="icon-link"
      href="https://ebay.com"
    >
      <svg
        aria-hidden="true"
        class="icon icon--settings"
        focusable="false"
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-settings"
        />
      </svg>
    </a>
  </p>
</DocumentFragment>
`;

exports[`Storyshots ebay-icon-button Transparent 1`] = `
<DocumentFragment>
  <p>
    <button
      aria-label="Menu"
      class="icon-btn icon-btn--transparent"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="icon icon--menu"
        focusable="false"
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-menu"
        />
      </svg>
    </button>
  </p>
</DocumentFragment>
`;

exports[`Storyshots ebay-icon-button With Badges 1`] = `
<DocumentFragment>
  <p>
    <button
      aria-label="Menu"
      class="icon-btn icon-btn--badged"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="icon icon--menu"
        focusable="false"
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-menu"
        />
      </svg>
      <span
        aria-hidden="true"
        aria-label="new feature available"
        class="badge"
      >
        1
      </span>
    </button>
  </p>
  <p>
    <a
      aria-label="Cart"
      class="icon-link icon-link--badged"
      href="https://ebay.com"
    >
      <svg
        aria-hidden="true"
        class="icon icon--cart"
        focusable="false"
        height="24px"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-cart"
        />
      </svg>
      <span
        aria-hidden="true"
        aria-label="3 items in your cart"
        class="badge"
      >
        3
      </span>
    </a>
  </p>
  <p>
    <a
      aria-label="Chat"
      class="icon-link icon-link--badged"
      href="https://ebay.com"
    >
      <svg
        aria-hidden="true"
        class="icon icon--chat-large"
        focusable="false"
        height="64px"
        width="64px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-chat-large"
        />
      </svg>
      <span
        aria-hidden="true"
        aria-label="99 unread messages"
        class="badge"
      >
        99
      </span>
    </a>
  </p>
</DocumentFragment>
`;
